src/ocean/util/container/btree/Implementation.d

```diff
@@ -0,0 +1,1121 @@
+/********************************************************************************
+
+    Internal (non-user facing) implementation of BTree.
+
+    Copyright:
+        Copyright (c) 2017 Sociomantic Labs GmbH.
+        All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.util.container.btree.Implementation;
+
+import ocean.transition;
+import ocean.core.array.Mutation;
+import ocean.core.array.Search;
+import ocean.core.Enforce;
+import ocean.core.Traits;
+import ocean.util.container.mem.MemManager;
+
+
+/// ditto
+package struct BTreeImplementation (ElementType, int tree_degree)
+{
+    /**************************************************************************
+
+        Root of the tree.
+
+    ***************************************************************************/
+
+    package BTreeNode* root;
+
+    /**************************************************************************
+
+        Node of the tree. Contains at most (degree * 2 - 1) elements and
+        (degree * 2) subtrees.
+
+    **************************************************************************/
+
+    package struct BTreeNode
+    {
+        /// Number of the elements currently in the node
+        int number_of_elements;
+
+        /// Indicator if the given node is a leaf
+        bool is_leaf;
+
+        /// Array of the elements
+        package Unqual!(ElementType)[tree_degree * 2 - 1] elements;
+
+        /// Array of the pointers to the subtrees
+        package BTreeNode*[tree_degree * 2] child_nodes;
+    }
+
+    /***************************************************************************
+
+        Type of the element actually stored in the node. The type stored in the
```

Comment:
Yeah good idea. The only way around it I could think is that we recreate this array on every change.